core/src/main/java/org/elasticsearch/index/percolator/PercolatorQueriesRegistry.java

```diff
@@ -179,7 +179,7 @@ private Query parseQuery(String type, XContentParser parser) {
         if (type != null) {
             previousTypes = QueryShardContext.setTypesWithPrevious(type);
         }
-        QueryShardContext context = queryParserService.getShardContext();
+        QueryShardContext context = queryShardContext.clone();
```

Comment:
can we avoid cloning the context here and maybe pass in some kind of small factory instead?